[docs] Add main documentation for Kubernetes Apps Discovery

AntonAM opened the pull request:
This PR adds main documentation for new feature - Kubernetes Apps Discovery. It also rearranges existing documentation a little bit - before new feature we already had Kubernetes Auto-Discovery, which was about discovery clusters. To align on terminology this PR amends name of existing feature in documentation to `Kubernetes Clusters Discovery`, while new feature is called `Kubernetes Apps Discovery`, so naming is clear about what the feature does.

References https://github.com/gravitational/teleport/pull/31037 (lint will fails until it's merged)

zmb3 commented on docs/pages/kubernetes-access/discovery-apps.mdx:
```diff
@@ -0,0 +1,182 @@
+---
+title: Kubernetes Apps Discovery
```
Is this the right title? The feature converts "Kubernetes Services" into "Teleport Apps." I fear introducing "Kubernetes Apps" will be confusing.

AntonAM replied:
Yep, naming was hard. First version of this documentation had "Kubernetes Services Discovery" 😅 
We also will probably have databases discovery inside Kube clusters in the future, so @tigrato suggested we call this one "Apps Discovery" and then add another section the docs side panel when we add databases, so we'll have something like

```
Kubernetes Apps Discovery
Kuberentes Databases Discovery
Kubernetes Clusters Discovery
  |- AWS
  |- Azure
  |- Google Cloud
```

Or combining Apps & Databases under "Services":
```
Kubernetes Services Discovery
|- Apps 
|- Database
Kubernetes Clusters Discovery
  |- AWS
  |- Azure
  |- Google Cloud
```

To be honest I'm not sure which way is better and open to suggestions.

zmb3 commented on docs/pages/kubernetes-access/discovery-apps.mdx:
```diff
@@ -0,0 +1,182 @@
+
+By default name of the created Teleport app will consist of Kubernetes service name, namespace and 
+Kubernetes cluster name: `$SERVICE_NAME-$NAMESPACE-$KUBE_CLUSTER_NAME`. That name can be changed by Kubernetes annotation 
+"teleport.dev/name".
```
```suggestion
`teleport.dev/name`.
```

To be consistent with above.

zmb3 reviewed (commented):
This is the first time I've read into the details of this feature. Looks super configurable - nice job with the annotations.

zmb3 commented:
Also note, _Kubernetes_ is spelled incorrectly in your commit message.

AntonAM commented on docs/pages/kubernetes-access/discovery-apps.mdx:
````diff
@@ -0,0 +1,182 @@
+
+```code
+$ helm install teleport-agent teleport/teleport-kube-agent \
+  --set roles=kube\,app\,discovery \
````
Chart generates config from template based on these roles, so it needs it in form of chart values, since token is just a opaque string for it.

AntonAM commented on docs/pages/kubernetes-access/discovery-apps.mdx:
```diff
@@ -0,0 +1,182 @@
+  enabled: yes
+  resources:
+    - labels:
+        "teleport.dev/kubernetes-cluster": "main-cluster"
```
I've changed it to `main-cluster` in the helm example.

ptgott commented on docs/config.json:
```diff
@@ -920,7 +920,11 @@
         },
         {
-          "title": "Kubernetes Discovery",
+          "title": "Kubernetes Apps Discovery",
```
```suggestion
          "title": "Kubernetes App Discovery",
```
Making this singular as in "Service Discovery" or "Incident Response"

AntonAM replied:
I did it on purpose, because we have a problem with name ambiguity when it singular. Now that we have both discovery **of** clusters and **inside** clusters, it's not clear which one "Kubernetes Cluster Discovery" means. So to make it more clear I used "Kubernetes Clusters Discovery" and using same pattern "Kubernetes Apps Discovery", though I don't think it's a hard requirement for the Apps, since we don't have this problem with ambiguity. Does it make sense?

ptgott replied:
I'm not sure I can gather the distinction between "of clusters" and "inside clusters" from the term "Kubernetes Clusters Discovery". To me, this term sounds like it means, "the discovery of Kubernetes clusters" in a way that clashes a bit with the conventions of written English (it should be "Kubernetes cluster discovery").

I would use another title to make this clearer.

AntonAM replied:
Yep, "Kubernetes Clusters Discovery" means "the discovery of Kubernetes clusters" and I think it's clear. The problem is with "Kubernetes Cluster Discovery" which to me is not as clear (it might means discovery inside cluster).

I'm fine with changing it to how you suggested if you think it doesn't sound confusing.
So we'll have:

"Kubernetes Cluster Discovery" - discovery of Kubernetes clusters (in AWS, GCP etc)
"Kubernetes App Discovery"      - discovery of Apps inside a Kubernetes cluster.

ptgott replied:
```
"Kubernetes Cluster Discovery" - discovery of Kubernetes clusters (in AWS, GCP etc)
"Kubernetes App Discovery" - discovery of Apps inside a Kubernetes cluster.
```
This is how I understand it

ptgott commented on docs/config.json:
```diff
@@ -920,7 +920,11 @@
+          "slug": "/kubernetes-access/discovery-apps/"
+        },
+        {
+          "title": "Kubernetes Clusters Discovery",
```
```suggestion
          "title": "Kubernetes Cluster Discovery",
```

ptgott commented on docs/pages/kubernetes-access/discovery-apps.mdx:
Should this page be in the "Application Access" section instead? This section of the docs is related to configuring/enabling access to Kubernetes clusters using Teleport.

ptgott reviewed (commented):
Left a couple of quick comments. I need to give this a more thorough review.

ptgott commented:
@AntonAM I'm going to suggest an alternative placement/structure for these docs in a PR against this branch!

ptgott commented:
@AntonAM Opened a PR here: https://github.com/gravitational/teleport/pull/31606

tigrato reviewed (approved):
@AntonAM can you please include the following recommendations https://github.com/gravitational/teleport/pull/31606#pullrequestreview-1623074786

I think it's quite essential to have a guide about how to upgrade from a non-app discovery agent to this agent with the discovery.

AntonAM commented:
@tigrato yep, I'll apply your suggestions from that PR, by merging I just moved the work here.

AntonAM committed:
Improve wording.

Co-authored-by: Zac Bergquist <zac.bergquist@goteleport.com>

AntonAM committed:
Add main documentation for Kubernetes Apps Discovery

AntonAM committed:
Improve wording.

Co-authored-by: Zac Bergquist <zac.bergquist@goteleport.com>

AntonAM committed:
Make example cluster names consistent

AntonAM committed:
Add internal link for annotation

AntonAM committed:
Remove unneeded symbol

Co-authored-by: Zac Bergquist <zac.bergquist@goteleport.com>

AntonAM committed:
Improve wording.

ptgott committed:
Restructure the Kubernetes app discovery guide (#31606)

- Move the docs to the Application Access section, since this material
  has to do with proxying applications via the Application Service.

- Split up the guide into separate guides for different documentation
  purposes: a how-to guide, configuration reference, and architecture
  guide.

- Fix linter/spelling issues

AntonAM committed:
Add not about running agent inside the Kuberentes cluster

AntonAM committed:
Add instruction how to enabled kube app discovery on existing installation

AntonAM committed:
Add example of configuration of teleport-kube-agent Helm chart

tigrato committed:
convert tip to tabs

tigrato committed:
fix wrong ref
